src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "lazy_tensor.hpp"
+
+using ov::npuw::weights::ConcatMeta;
+using ov::npuw::weights::LazyTensor;
+using ov::npuw::weights::Transform;
+using ov::npuw::weights::TransformType;
+
+std::size_t LazyTensor::Hash::operator()(const LazyTensor& lt) const {
+    std::size_t seed = std::hash<void*>()(lt.m_orig_data) + 0x9e3779b9;
+    seed ^= std::hash<std::string>()(lt.m_orig_shape.to_string()) + 0x9e3779b9;
+    seed ^= std::hash<std::string>()(lt.m_orig_type.to_string()) + 0x9e3779b9;
+    for (const auto& tr : lt.m_transforms) {
+        seed ^= std::hash<int>()(static_cast<int>(tr.first)) + 0x9e3779b9;
+        if (tr.first == TransformType::PERMUTE) {
+            const auto& axes = std::get<std::vector<std::size_t>>(tr.second);
+            for (const auto& axis : axes) {
+                seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+            }
+        } else if (tr.first == TransformType::CONCAT) {
+            // concat tag can be different, no need to hash it
+            const auto& axis = std::get<ConcatMeta>(tr.second).second;
+            seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+        }
+    }
+    return seed;
+}
+
+LazyTensor::LazyTensor(const TransformType& type, const Transform& transform) {
+    // Sanity check
+    NPUW_ASSERT(type == TransformType::TENSOR && std::holds_alternative<ov::Tensor>(transform));
+    m_transforms.push_back({type, transform});
+    const auto& tensor = std::get<ov::Tensor>(transform);
+    m_orig_data = tensor.data();
+    m_orig_shape = tensor.get_shape();
+    m_orig_type = tensor.get_element_type();
+}
+
+bool LazyTensor::operator==(const LazyTensor& other) const {
+    if (m_orig_data != other.m_orig_data || m_orig_shape != other.m_orig_shape || m_orig_type != other.m_orig_type ||
+        m_transforms.size() != other.m_transforms.size()) {
+        return false;
+    }
+
+    for (size_t i = 0; i < m_transforms.size(); ++i) {
+        if (m_transforms[i].first != other.m_transforms[i].first) {
+            return false;
+        }
+        // Only PERMUTE and CONCAT have meta which needs to be compared
+        if (m_transforms[i].first == TransformType::PERMUTE) {
+            if (std::get<std::vector<std::size_t>>(m_transforms[i].second) !=
+                std::get<std::vector<std::size_t>>(other.m_transforms[i].second)) {
+                return false;
+            }
+        } else if (m_transforms[i].first == TransformType::CONCAT) {
+            // concat tag can be different, no need to compare it
+            if (std::get<ConcatMeta>(m_transforms[i].second).second !=
+                std::get<ConcatMeta>(other.m_transforms[i].second).second) {
+                return false;
+            }
+        }
+    }
```

Comment:
as you (should) have an up-to-date hash, it could be just a hash comparison here.

Reply:
For the exact match I still need to compare some fields with operator==